File: apple/InlineKit/Sources/RealtimeV2/Transport/MockTransport.swift
import AsyncAlgorithms
import Foundation
import InlineProtocol

/// Pure in-memory implementation that lets unit tests feed pre-canned events
/// and observe outbound messages without touching the network stack.
actor MockTransport: Transport {
  // MARK: Public -----------------------------------------------------------------

  nonisolated var events: AsyncChannel<TransportEvent> { channel }

  func start() async {
    guard !started else { return }
    started = true
    await channel.send(.connecting)
    await channel.send(.connected)
  }

  func stop() async {
    guard started else { return }
    started = false
  }

  func send(_ message: ClientMessage) async throws {
    sentMessages.append(message)
  }

  // MARK: Testing helpers --------------------------------------------------------

  /// Push an arbitrary event into the stream (used by unit tests).
  func emit(_ event: TransportEvent) async {
    await channel.send(event)
  }

  /// All messages that have been sent by the system under test.
  private(set) var sentMessages: [ClientMessage] = []

  // MARK: Private ----------------------------------------------------------------

  private var started = false

  private let channel: AsyncChannel<TransportEvent>

  init() {
    channel = AsyncChannel<TransportEvent>()
  }
}

File: web/packages/client/src/realtime/transport/mock-transport.ts
import type { ClientMessage, ServerProtocolMessage } from "@in/protocol/core"
import { AsyncChannel } from "../../utils/async-channel"
import type { TransportEvent } from "../types"
import { TransportError, type Transport } from "./transport"

export type MockTransportState = "idle" | "connecting" | "connected"

export class MockTransport implements Transport {
  readonly events = new AsyncChannel<TransportEvent>()
  readonly sent: ClientMessage[] = []

  state: MockTransportState = "idle"

  async start() {
    if (this.state !== "idle") return
    this.state = "connecting"
    await this.events.send({ type: "connecting" })
  }

  async stop() {
    if (this.state === "idle") return
    this.state = "idle"
    await this.events.send({ type: "stopping" })
  }

  async send(message: ClientMessage) {
    if (this.state !== "connected") {
      throw TransportError.notConnected()
    }
    this.sent.push(message)
  }

  async stopConnection() {
    if (this.state === "idle") return
    this.state = "connecting"
  }

  async reconnect() {
    this.state = "connecting"
    await this.events.send({ type: "connecting" })
  }

  async connect() {
    this.state = "connected"
    await this.events.send({ type: "connected" })
  }

  async emitMessage(message: ServerProtocolMessage) {
    await this.events.send({ type: "message", message })
  }
}
